ASoC: Intel: Add machine driver for da7219_max98373 on ICL/JSL platforms

yongzhi1 opened the pull request:
This patch add jsl_da7219_max98373 machine driver.

Tested on IceLake and JasperLake platforms with SOF only.

Signed-off-by: Yong Zhi <yong.zhi@intel.com>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>
Signed-off-by: Vani Ganji <vani.ganji@intel.com>

plbossart commented on sound/soc/intel/boards/jsl_da7219_max98373.c:
```diff
@@ -0,0 +1,421 @@
+#define SOF_DA7219_MCLK_24MHZ	BIT(0)
+
+/* Default: MCLK 24M */
+static unsigned long sof_da7219_quirk = SOF_DA7219_MCLK_24MHZ;
```
there is no way this is the case for ICL (38.4)
I would think JSL is the same as GLK (19.2)?

yongzhi1 replied:
The clk was confirmed with Kar before.

plbossart commented on sound/soc/intel/boards/jsl_da7219_max98373.c:
```diff
@@ -0,0 +1,421 @@
+			}
+		}
+		if (!strcmp(codec_dai->component->name, MAXIM_DEV1_NAME)) {
+			ret = snd_soc_dai_set_tdm_slot(codec_dai, 0xC0, 3, 8, 16);
```
can you add comments to clarify how the traffic is split?
looks like playback is on slots 0/1 and capture on either slots 4/5 (dev0) or 6/7 (dev1)?

yongzhi1 replied:
That's correct, same as kbl_da7219_max98373.

plbossart commented on sound/soc/intel/boards/jsl_da7219_max98373.c:
```diff
@@ -0,0 +1,421 @@
+		return -EINVAL;
+
+	return hdac_hdmi_jack_port_init(component, &card->dapm);
+}
```
you need to redo all this to handle the new HDMI common parts

plbossart commented on sound/soc/intel/boards/jsl_da7219_max98373.c:
```diff
@@ -0,0 +1,421 @@
+		.no_pcm = 1,
+		.dai_fmt = SND_SOC_DAIFMT_DSP_B |
+			SND_SOC_DAIFMT_NB_NF |
+			SND_SOC_DAIFMT_CBS_CFS,
```
topology defined, should be removed.

yongzhi1 replied:
Ack.

plbossart commented on sound/soc/intel/boards/jsl_da7219_max98373.c:
```diff
@@ -0,0 +1,421 @@
+		.dpcm_playback = 1,
+		.dpcm_capture = 1, /* IV feedback */
+		.ops = &ssp1_ops,
+		SND_SOC_DAILINK_REG(ssp1_pin, ssp1_codec, platform),
```
ssp1_amps would be clearer.

yongzhi1 replied:
Ack.

plbossart commented on sound/soc/intel/boards/jsl_da7219_max98373.c:
```diff
@@ -0,0 +1,421 @@
+		.dpcm_capture = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(dmic_pin, dmic_codec, platform),
+	},
```
dmic16k?

yongzhi1 replied:
Ack.

plbossart commented on sound/soc/intel/common/soc-acpi-intel-icl-match.c:
```diff
@@ -29,6 +28,13 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_icl_machines[] = {
 	},
+	{
+		.id = "DLGS7219",
+		.drv_name = "jsl_da7219_max98373",
```
can't you play with aliases and use icl_da7219_max...

yongzhi1 replied:
Ack.

plbossart commented on sound/soc/intel/common/soc-acpi-intel-icl-match.c:
```diff
@@ -29,6 +28,13 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_icl_machines[] = {
+		.drv_name = "jsl_da7219_max98373",
+		.machine_quirk = snd_soc_acpi_codec_list,
+		.sof_fw_filename = "sof-icl.ri",
+		.sof_tplg_filename = "sof-jsl-da7219.tplg",
```
sof-icl, and create the relevant file from the same template.

yongzhi1 replied:
Ack.

plbossart commented on sound/soc/intel/boards/Kconfig:
```diff
@@ -498,4 +498,22 @@ config SND_SOC_INTEL_TGL_RT1308_MACH
+if (SND_SOC_SOF_JASPERLAKE || SND_SOC_SOF_ICELAKE) && SND_SOC_SOF_HDA_LINK
+
+config SND_SOC_INTEL_JSL_DA7219_MAX98373_MACH
+	tristate "JSL with DA7219 and MAX98373 in I2S Mode"
```
i'd like to have two Kconfigs that use the same common parts, see e.g. how it's done for bxt_da7219_mx98753a

yongzhi1 replied:
Ack

plbossart commented on sound/soc/intel/boards/jsl_da7219_max98373.c:
```diff
@@ -0,0 +1,421 @@
+		.driver_data = (void *)(0),
+	},
+	{}
+};
```
i don't think you really need this if you use the board id and use different aliases for jsl and icl drivers.

yongzhi1 replied:
MCLK is different for icl and jsl, so the quirk is used to program da7219 with different freq, but it may make more sense to support jsl alone for now and add icl support when future needs arise.

plbossart commented:
@yongzhi1 any updates? this has been parked for 14 days now

kv2019i commented on sound/soc/intel/boards/Kconfig:
```diff
@@ -498,4 +498,22 @@ config SND_SOC_INTEL_TGL_RT1308_MACH
+	select SND_SOC_DA7219
+	select SND_SOC_MAX98373
+	select SND_SOC_DMIC
+	select SND_SOC_HDAC_HDMI
```
@yongzhi1 hdac-hdmi support for ICL and JSL is dropped from upstream, so you need to you snd-hda-codec-hdmi. Please use

select SND_HDA_CODEC_HDMI if SND_SOC_SOF_HDA_CODEC

here.

yongzhi1 replied:
Ack, will switch to SND_HDA_CODEC_HDMI.

kv2019i commented on sound/soc/intel/boards/jsl_da7219_max98373.c:
```diff
@@ -0,0 +1,421 @@
+static int platform_clock_control(struct snd_soc_dapm_widget *w,
+	struct snd_kcontrol *k, int  event)
+{
+	int ret = 0;
```
reverse-xmas tree introduction of variables is preferered (so "int ret = 0" as last as it's the shortest).

yongzhi1 replied:
Sure, will apply the principle to similar places.

kv2019i commented on sound/soc/intel/boards/jsl_da7219_max98373.c:
```diff
@@ -0,0 +1,421 @@
+static struct snd_soc_jack jsl_headset;
+static int jsl_da7219_codec_init(struct snd_soc_pcm_runtime *rtd)
+{
+	int ret;
```
Reverse-xmas tree would be preferred here as well.

yongzhi1 replied:
Ack.

kv2019i commented on sound/soc/intel/boards/jsl_da7219_max98373.c:
```diff
@@ -0,0 +1,421 @@
+	char jack_name[32];
+	unsigned int i = 0;
+	int err;
+
```
Please add support for snd-hda-codec-hdmi here. E.g. look at late_probe() of sof_rt5682.c (commit "ASoC: intel: sof_rt5682: common hdmi codec support" is an example).

yongzhi1 replied:
Ack. my understanding is that we can remove SND_HDA_CODEC_HDMI related code since it's be dropped for new platform.

yongzhi1 commented:
Updated the patch at https://github.com/yongzhi1/linux-1/commit/995404024eb630db5c5a04210d206a6ce2c01674

Main changes:
-  Switched to HDA hdmi codec.
-  Removed quirks for code clarity, ICL support can be added in the future if need arises.
-  Addressed review comments on coding style etc.

kv2019i commented on sound/soc/intel/boards/Kconfig:
```diff
@@ -498,4 +498,22 @@ config SND_SOC_INTEL_TGL_RT1308_MACH
+	select SND_SOC_DA7219
+	select SND_SOC_MAX98373
+	select SND_SOC_DMIC
+	select SND_SOC_HDAC_HDMI if SND_SOC_SOF_HDA_LINK
```
@yongzhi1 oops, line510 should be replaced with:
select SND_HDA_CODEC_HDMI if SND_SOC_SOF_HDA_CODEC

yongzhi1 replied:
Thanks for catching it, updated https://github.com/yongzhi1/linux-1/commit/fd2cfb741c28a129f583b0f9065e8d7e01739467

plbossart commented on sound/soc/intel/boards/Kconfig:
```diff
@@ -498,4 +498,22 @@ config SND_SOC_INTEL_TGL_RT1308_MACH
+	select SND_SOC_DA7219
+	select SND_SOC_MAX98373
+	select SND_SOC_DMIC
+	select SND_HDA_CODEC_HDMI if SND_SOC_SOF_HDA_LINK
```
@yongzhi1 is this correct? all other Kconfigs use:

	select SND_HDA_CODEC_HDMI if SND_SOC_SOF_HDA_CODEC

@kv2019i can you help?

kv2019i replied:
@plbossart @yongzhi1 yes, HDA_CODEC is the correct one, HDA_LINK doesn't bring all the dependencies in, i.e.
select SND_HDA_CODEC_HDMI if SND_SOC_SOF_HDA_CODEC

yongzhi1 replied:
I did not see SND_SOC_SOF_HDA_CODEC used in elsewhere other than this file, where is this config introduced?

plbossart replied:
@kv2019i @yongzhi1 is right, something's off in the Kconfigs

````
$ git grep SND_SOC_SOF_HDA_CODEC
sound/soc/intel/boards/Kconfig: select SND_HDA_CODEC_HDMI if SND_SOC_SOF_HDA_CODEC
sound/soc/intel/boards/Kconfig: select SND_HDA_CODEC_HDMI if SND_SOC_SOF_HDA_CODEC
sound/soc/intel/boards/Kconfig: select SND_HDA_CODEC_HDMI if SND_SOC_SOF_HDA_CODEC
sound/soc/intel/boards/Kconfig: select SND_HDA_CODEC_HDMI if SND_SOC_SOF_HDA_CODEC
sound/soc/intel/boards/Kconfig: select SND_HDA_CODEC_HDMI if SND_SOC_SOF_HDA_CODEC
sound/soc/intel/boards/Kconfig: select SND_HDA_CODEC_HDMI if SND_SOC_SOF_HDA_CODEC
sound/soc/intel/boards/Kconfig: select SND_HDA_CODEC_HDMI if SND_SOC_SOF_HDA_CODEC
````

plbossart replied:
was it supposed to be  SND_SOC_SOF_HDA_AUDIO_CODEC ?

kv2019i replied:
@plbossart Right. Thanks @yongzhi1 you are right. Not sure how this slipped through, but it's indeed wrong. It's indeed  SND_SOC_SOF_HDA_AUDIO_CODEC . I'll make a patch for the other machine drivers.

kv2019i replied:
-> https://github.com/thesofproject/linux/pull/1611

plbossart commented on sound/soc/intel/boards/jsl_da7219_max98373.c:
```diff
@@ -0,0 +1,370 @@
+#define MAXIM_DEV0_NAME		"i2c-MX98373:00"
+#define MAXIM_DEV1_NAME		"i2c-MX98373:01"
+
+struct jsl_hdmi_pcm {
```
remove the jsl prefix, it's not needed and will prevent from reusing this machine driver for other machines. The more common the structures are the easier it is to maintain.

yongzhi1 replied:
Should I rename the file name to sof_da7219_max98373.c?

plbossart replied:
yes, that would probably be a good thing.

plbossart commented on sound/soc/intel/boards/jsl_da7219_max98373.c:
```diff
@@ -0,0 +1,370 @@
+};
+
+struct jsl_card_private {
+	struct snd_soc_jack jsl_headset;
```
remove those two prefixes

yongzhi1 replied:
Ack

plbossart commented on sound/soc/intel/boards/jsl_da7219_max98373.c:
```diff
@@ -0,0 +1,370 @@
+struct jsl_card_private {
+	struct snd_soc_jack jsl_headset;
+	struct list_head hdmi_pcm_list;
+	struct snd_soc_jack jsl_hdmi[3];
```
and this prefix as well

yongzhi1 replied:
Ack

plbossart commented on sound/soc/intel/boards/jsl_da7219_max98373.c:
```diff
@@ -0,0 +1,370 @@
+};
+
+static struct snd_soc_jack jsl_headset;
+static int jsl_da7219_codec_init(struct snd_soc_pcm_runtime *rtd)
```
remove jsl prefixes

yongzhi1 replied:
Ack

plbossart commented on sound/soc/intel/boards/jsl_da7219_max98373.c:
```diff
@@ -0,0 +1,370 @@
+	return ret;
+}
+
+static int jsl_ssp1_hw_params(struct snd_pcm_substream *substream,
```
remove prefix

yongzhi1 replied:
Ack

plbossart commented on sound/soc/intel/boards/jsl_da7219_max98373.c:
```diff
@@ -0,0 +1,370 @@
+		}
+		if (!strcmp(codec_dai->component->name, MAXIM_DEV1_NAME)) {
+			/* vmon_slot_no = 6 imon_slot_no = 7 for TX slots */
+			ret = snd_soc_dai_set_tdm_slot(codec_dai, 0xC0, 3, 8, 16);
```
btw is there a reason why we use 8 slots if you only need 4?

yongzhi1 replied:
My understanding 8 is the total slots, active slots can be derived from rx/tx mask, I could be wrong here, will take a look.

plbossart replied:
you have separate lines for RX and TX, so what matters is the maximum number of slots used for the same direction. Since the SSP is symmetric, you need to use that number for RX and TX and then round (ceil) to the nearest power of two.

yongzhi1 replied:
Separate calls for snd_soc_dai_set_tdm_slot() are for left and right speakers, RX uses either 4/5 or 6/7, TX uses 1/2, max number for both RX and TX are 2, @slots is the total number of slots of the TDM stream, do you mean to change this value to 2 or 4?

plbossart replied:
on RX you use 4 slots total, so if I get this right you could use 4 slots of the TDM bus.
Is there a specific reason why you chose 8?

yongzhi1 replied:
The reason for 8 is that I think by default *tx_mask = (1 << slots) - 1;
https://elixir.bootlin.com/linux/v5.5-rc1/source/sound/soc/soc-dai.c#L123
so seems mask width cannot be greater than the slots?

plbossart replied:
if you used 4 slots, you'd need to also change the allocation for each channel, of course you cannot use slot 4..7 if you only have 4 slots...

yongzhi1 replied:
Ack, will update to 4 slots based on suggestion.

plbossart commented on sound/soc/intel/boards/jsl_da7219_max98373.c:
```diff
@@ -0,0 +1,370 @@
+	},
+};
+
+static int jsl_hdmi_init(struct snd_soc_pcm_runtime *rtd)
```
remove prefix

yongzhi1 replied:
Ack

plbossart commented on sound/soc/intel/boards/jsl_da7219_max98373.c:
```diff
@@ -0,0 +1,370 @@
+	return 0;
+}
+
+static int jsl_card_late_probe(struct snd_soc_card *card)
```
remove jsl prefix

yongzhi1 replied:
Ack

plbossart commented on sound/soc/intel/common/soc-acpi-intel-jsl-match.c:
```diff
@@ -10,6 +10,13 @@
+	{
+		.id = "DLGS7219",
+		.drv_name = "jsl_da7219_max98373",
+		.machine_quirk = snd_soc_acpi_codec_list,
```
is this needed if you don't have a .quirk_data?

yongzhi1 replied:
Ack, will check.

yongzhi1 commented:
v3 https://github.com/yongzhi1/linux-1/commit/e626d7a8ff2c37fef91af05bddf3144fe9f19473 update:
1. Changed spk amp tdm slots from 8 to 4.
2. Removed jsl prefix at various places.
3. Updated Kconfig to use SND_SOC_SOF_HDA_AUDIO_CODEC

plbossart commented on sound/soc/intel/boards/sof_da7219_max98373.c:
```diff
@@ -0,0 +1,370 @@
+	{ "Headset Mic", NULL, "Platform Clock" },
+};
+
+static struct snd_soc_jack headset;
```
newline here?

yongzhi1 replied:
Ack, added new line https://github.com/yongzhi1/linux-1/commit/110f33b3f6bba9d91feadcba175cf216ce749aba, thanks for review!!

sathya-nujella commented on sound/soc/intel/boards/Kconfig:
```diff
@@ -498,4 +498,22 @@ config SND_SOC_INTEL_TGL_RT1308_MACH
 
+if SND_SOC_SOF_JASPERLAKE
+
+config SND_SOC_INTEL_JSL_DA7219_MAX98373_MACH
```
machine driver name is generic. 
1) but CONFIG is specific to JSL? can this be generic like... e.g. SND_SOC_INTEL_SOF_RT5682_MACH...
2) Also if generic, can this be moved out of SND_SOC_SOF_JASPERLAKE scope?

yongzhi1 replied:
I am OK to make it the first platform-blind machine driver if no different opinion here.

plbossart replied:
its' not the first one, we already have sof_rt5682.

yongzhi1 replied:
I mean sof_rt5682 is still conditionally configured by SND_SOC_SOF_BAYTRAIL.

plbossart replied:
well here as well, we only want this exposed with JSL for now.

sathya-nujella commented on sound/soc/intel/boards/sof_da7219_max98373.c:
```diff
@@ -0,0 +1,371 @@
+static struct platform_driver audio = {
+	.probe = audio_probe,
+	.driver = {
+		.name = "jsl_da7219_max98373",
```
can this also be generic please? may be sof_da7219_max98373 ?

yongzhi1 replied:
Ack. will also update line 307 from jslda7219max to da7219max.

yongzhi1 commented:
Thanks @plbossart and @sathya-nujella for v3 review.
v4 update https://github.com/yongzhi1/linux-1/commit/f005ee0daec226f46a7d127dd0fe2a2b70b437bd

- removed ref to jsl at driver name and Kconfig file.

plbossart commented on sound/soc/intel/boards/sof_da7219_max98373.c:
```diff
@@ -0,0 +1,371 @@
+		.name = "sof_da7219_max98373",
+		.pm = &snd_soc_pm_ops,
+	},
+	.id_table = jsl_board_ids,
```
@yongzhi1 now that you've removed most prefixes, could we do one more update and remove this last one?

yongzhi1 replied:
Ack, updated https://github.com/yongzhi1/sof/commit/b36673c217a2468046470affa5a0a4c1ea99bb33

yongzhi1 replied:
Sorry wrong link, should be https://github.com/yongzhi1/linux-1/commit/a15d9f3299710aa3360a4b61407bdb1976ff9a8e

lyakh commented:
> please don't add jsl_ prefixes, it was a bad habit that we are trying to suppress to make sharing code easier.
> couple of other nit picks below

@plbossart I understand the "code sharing" argument, but there's also at least one argument in favour of using prefixes: code tagging. I use emacs but this would also be the case in vi, I think. With prefixes your symbol names are unique, so, when you follow such a symbol you quickly get to the correct code. However with no prefixes several files can have symbols with the same name, then getting to the right one becomes more difficult. \<rant\>In some cases this name re-use makes using tags unfeasible, so you have to revert to grepping \<\/rant\>
Actually there's also a second argument: if some functions in a file are global and some are local. For global functions you have to use a namespace prefix to distinguish them from other namespaces. But then using a prefix on global functions and not using one on static ones doesn't look good.

plbossart commented:
> > please don't add jsl_ prefixes, it was a bad habit that we are trying to suppress to make sharing code easier.
> > couple of other nit picks below
> 
> @plbossart I understand the "code sharing" argument, but there's also at least one argument in favour of using prefixes: code tagging. I use emacs but this would also be the case in vi, I think. With prefixes your symbol names are unique, so, when you follow such a symbol you quickly get to the correct code. However with no prefixes several files can have symbols with the same name, then getting to the right one becomes more difficult. <rant>In some cases this name re-use makes using tags unfeasible, so you have to revert to grepping </rant>
> Actually there's also a second argument: if some functions in a file are global and some are local. For global functions you have to use a namespace prefix to distinguish them from other namespaces. But then using a prefix on global functions and not using one on static ones doesn't look good.

I personally find it silly to use prefixes on static functions.
